### PR TITLE
Unsubscribe

### PR DESCRIPTION
**Pull request by cindyyan317**

### CODE REVIEW

File: src/rpc/ngHandlers/Subscribe.h [hunk 1]
```diff
@@ -71,11 +71,42 @@ class BaseSubscribeHandler
...
+            validation::CustomValidator{[](boost::json::value const& value, std::string_view key) -> MaybeError {
+                if (!value.is_array())
+                    return Error{RPC::Status{RPC::RippledError::rpcINVALID_PARAMS, std::string(key) + "NotArray"}};
+                for (auto const& book : value.as_array())
+                {
+                    if (!book.is_object())
+                        return Error{
+                            RPC::Status{RPC::RippledError::rpcINVALID_PARAMS, std::string(key) + "ItemNotObject"}};
+                    if (book.as_object().contains("both") && !book.as_object().at("both").is_bool())
+                    {
```
Comment (godexsoft) [inline src/rpc/ngHandlers/Subscribe.h]: May as well remove the extra scope here same as on the line above as there is no complexity to the statement. same for below if statements :)

File: src/rpc/ngHandlers/Unsubscribe.h [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+            validation::CustomValidator{[](boost::json::value const& value, std::string_view key) -> MaybeError {
+                if (!value.is_array())
+                    return Error{RPC::Status{RPC::RippledError::rpcINVALID_PARAMS, std::string(key) + "NotArray"}};
+                for (auto const& book : value.as_array())
+                {
+                    if (!book.is_object())
+                        return Error{
+                            RPC::Status{RPC::RippledError::rpcINVALID_PARAMS, std::string(key) + "ItemNotObject"}};
+                    if (book.as_object().contains("both") && !book.as_object().at("both").is_bool())
+                    {
```
Comment (godexsoft) [inline src/rpc/ngHandlers/Unsubscribe.h]: same as above - probably no need for the extra braces here

File: src/rpc/ngHandlers/Unsubscribe.h [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+            {JS(books), booksValidator}};
+
+        return rpcSpec;
+    }
+
+    Result
+    process(Input input, Context const& ctx) const
+    {
+        if (input.streams)
+        {
```
Comment (godexsoft) [inline src/rpc/ngHandlers/Unsubscribe.h]: I'd omit braces here too, especially now that we have more characters per column and it won't look bad if you omit the braces

File: src/rpc/ngHandlers/Unsubscribe.h [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+        if (input.books)
+        {
+            unsubscribeToBooks(*(input.books), ctx.session);
+        }
+        return Output{};
+    }
+
+private:
+    void
+    unsubscribeToStreams(std::vector<std::string> const& streams, std::shared_ptr<WsBase> const& session) const
```
Comment (godexsoft) [inline src/rpc/ngHandlers/Unsubscribe.h]: it's probably old code that you copied but consider renaming to something like 'unsubscribeFromStreams' :)

File: src/rpc/ngHandlers/Unsubscribe.h [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+    {
+        for (auto const& s : streams)
```
Comment (godexsoft) [inline src/rpc/ngHandlers/Unsubscribe.h]: perhaps 'stream' is a better name here

File: src/rpc/ngHandlers/Unsubscribe.h [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+    {
+        for (auto const& account : accounts)
+        {
+            auto accountID = RPC::accountFromStringStrict(account);
+            subscriptions_->unsubAccount(*accountID, session);
+        }
+    }
+
+    void
+    unsubscribeToProposedAccounts(std::vector<std::string> accountsProposed, std::shared_ptr<WsBase> const& session)
```
Comment (godexsoft) [inline src/rpc/ngHandlers/Unsubscribe.h]: Probably should be 'From' here and below too

File: unittests/rpc/handlers/UnsubscribeTest.cpp [hunk 7]
```diff
@@ -0,0 +1,640 @@
...
+        auto const output = handler.process(input, Context{std::ref(yield), session_});
+        ASSERT_TRUE(output);
+        EXPECT_TRUE(output->as_object().empty());
+    });
+}
+
+TEST_F(RPCUnsubscribeTest, Books)
+{
+    auto const input = json::parse(fmt::format(
+        R"({{
```
Comment (godexsoft) [inline unittests/rpc/handlers/UnsubscribeTest.cpp:573]: Formatting of JSON again 😄 here and below 🙏

**Review by godexsoft — CHANGES_REQUESTED** [review]
Leaving small suggestions - nothing major 🙇‍♂️

File: src/rpc/ngHandlers/Unsubscribe.h (hunk 4 above)
Comment (cindyyan317) [reply src/rpc/ngHandlers/Unsubscribe.h]: yea, 'from' makes more sense

**Review by godexsoft — APPROVED** [review]
LGTM 🔥

**Pull request merged**